pkg/controllers/rollout/controller_test.go

```diff
@@ -388,6 +395,80 @@ func TestPickBindingsToRoll(t *testing.T) {
 			tobeUpdatedBindings:        []int{0, 1, 2},
 			needRoll:                   true,
 		},
+		"test bound with failed to apply bindings when there is no max surge allowed": {
+			allBindings: []*fleetv1beta1.ClusterResourceBinding{
+				generateFailedToApplyClusterResourceBinding(fleetv1beta1.BindingStateBound, "snapshot-1", cluster1),
+				generateFailedToApplyClusterResourceBinding(fleetv1beta1.BindingStateBound, "snapshot-1", cluster2),
+				generateClusterResourceBinding(fleetv1beta1.BindingStateBound, "snapshot-1", cluster3),
+				generateClusterResourceBinding(fleetv1beta1.BindingStateBound, "snapshot-1", cluster4),
+				generateClusterResourceBinding(fleetv1beta1.BindingStateBound, "snapshot-1", cluster5),
+			},
+			latestResourceSnapshotName: "snapshot-2",
+			crp:                        noMaxSurgeCRP,
+			tobeUpdatedBindings:        []int{0, 1},
+			needRoll:                   true,
+		},
+		"test unscheduled bindings": {
+			allBindings: []*fleetv1beta1.ClusterResourceBinding{
+				generateClusterResourceBinding(fleetv1beta1.BindingStateUnscheduled, "snapshot-1", cluster1),
+				generateClusterResourceBinding(fleetv1beta1.BindingStateUnscheduled, "snapshot-1", cluster2),
+			},
+			latestResourceSnapshotName: "snapshot-2",
+			crp: clusterResourcePlacementForTest("test",
+				createPlacementPolicyForTest(fleetv1beta1.PickNPlacementType, 2)),
+			tobeUpdatedBindings: []int{},
+			needRoll:            true,
+		},
+		"test with no bindings": {
+			allBindings:                []*fleetv1beta1.ClusterResourceBinding{},
+			latestResourceSnapshotName: "snapshot-2",
+			crp: clusterResourcePlacementForTest("test",
+				createPlacementPolicyForTest(fleetv1beta1.PickNPlacementType, 5)),
+			tobeUpdatedBindings: []int{},
+			needRoll:            false,
+		},
+		"test with failed unscheduled bindings": {
```

Comment:
there is no such thing as a failed to apply unscheduled binding. We should test unscheduled binding get deleted instead.